aten/src/ATen/native/native_functions.yaml

```diff
@@ -8323,29 +8321,27 @@
 
 - func: upsample_nearest1d.out(Tensor self, int[1] output_size, float? scales=None, *, Tensor(a!) out) -> Tensor(a!)
   python_module: nn
+  structured: True
   dispatch:
     CPU: upsample_nearest1d_out_cpu
     CUDA: upsample_nearest1d_out_cuda
 
 - func: upsample_nearest1d(Tensor self, int[1] output_size, float? scales=None) -> Tensor
   use_c10_dispatcher: full
   python_module: nn
-  dispatch:
-    CPU: upsample_nearest1d_cpu
-    CUDA: upsample_nearest1d_cuda
+  abstract: True  # HACK
```

Comment:
I don't intend to land the PR with the abstract keyword. Need to do some major pre-surgery to get rid though.

> But I wonder if we won't run into trouble when the naming convention needs to be disrupted for one reason or other.

There's no overload naming convention here; matching of signatures is done purely by looking at the type of the function, and the function name (sans overload name). That being said...

> In that world we might want a field that actually specifies the structured op we want to delegate to - delegate: upsample_nearest_1d.out or something.

I'm OK with doing this. It is technically redundant information but it gives us an opportunity to give better error messages and makes the intent in YAML more clear. That being said...

> opposed to silently generating code that would result in link errors 20 minutes later or whatever :)

I'm pretty sure that in the current formulation if you mismatch the signatures of two functions in the same structured block, you'll get a codegen error (because the dangling signature will fail to find the _out version).

Reply:
You're absolutely right, the grouping machinery is definitely robust enough to handle all the cases I was thinking about (missing `out` target, multiple clusters in a single op, overload name shenanigans). So I think the only motivation that remains is the thing about improving errors based on their expressed intent.

Reply:
@bhosmer There are also more serious changes we can make to how we setup `native_functions.yaml`; for example, we can force the user to group related signatures. This actually hearkens to my original original proposal (https://fb.quip.com/Z7PYApb0uqwR#VBbACAwaTQ4 FB-only sorry), but people seemed allergic to it the first time around.

Reply:
> people seemed allergic to it the first time around.

Do you remember what the counterarguments were? I always thought grouping variants together was a good idea - not just for readability in the YAML but also to spare the intake logic from performing exactly that grouping logic (plus sanity checks etc.) behind the scenes. 

Now that structured kernels will be adding an explicit intra-group relationship, having a declaration syntax that puts members together makes even more sense to me. 

I guess one possible issue is if different variants support different dispatch keys. Does that ever happen?

Reply:
For the record, the new version of the patch doesn't group `native_functions.yaml`, but has `structured` and `structured_delegate` to make the structure here more explicit